source/math/FloatMatrix.ooc

```diff
@@ -293,48 +293,90 @@ FloatMatrix: cover {
-		result[1, 0] = - (t[0, 1] * t[2, 2] - t[2, 1] * t[0, 2])
-		result[2, 0] = t[0, 1] * t[1, 2] - t[1, 1] * t[0, 2]
-		result[0, 1] = - (t[1, 0] * t[2, 2] - t[2, 0] * t[1, 2])
-		result[1, 1] = t[0, 0] * t[2, 2] - t[2, 0] * t[0, 2]
-		result[2, 1] = - (t[0, 0] * t[1, 2] - t[1, 0] * t[0, 2])
-		result[0, 2] = t[1, 0] * t[2, 1] - t[2, 0] * t[1, 1]
-		result[1, 2] = - (t[0, 0] * t[2, 1] - t[2, 0] * t[0, 1])
-		result[2, 2] = t[0, 0] * t[1, 1] - t[1, 0] * t[0, 1]
+		matrixOrder := t order
+		if (matrixOrder == 1)
+			result[0, 0] = t[0, 0]
+		else if (matrixOrder == 2)
+			result[0, 0] = t[0, 0] * t[1, 1] - t[0, 1] * t[1, 0]
+		else if (matrixOrder == 3) {
+			result[0, 0] = t[1, 1] * t[2, 2] - t[2, 1] * t[1, 2]
+			result[1, 0] = - (t[0, 1] * t[2, 2] - t[2, 1] * t[0, 2])
+			result[2, 0] = t[0, 1] * t[1, 2] - t[1, 1] * t[0, 2]
+			result[0, 1] = - (t[1, 0] * t[2, 2] - t[2, 0] * t[1, 2])
+			result[1, 1] = t[0, 0] * t[2, 2] - t[2, 0] * t[0, 2]
+			result[2, 1] = - (t[0, 0] * t[1, 2] - t[1, 0] * t[0, 2])
+			result[0, 2] = t[1, 0] * t[2, 1] - t[2, 0] * t[1, 1]
+			result[1, 2] = - (t[0, 0] * t[2, 1] - t[2, 0] * t[0, 1])
+			result[2, 2] = t[0, 0] * t[1, 1] - t[1, 0] * t[0, 1]
+		} else {
+			for (row in 0 .. t height) {
+				sign := (row % 2 == 0) ? 1.f : -1.f
+				for (col in 0 .. t width) {
+					result[row, col] = t _assembleWithoutRowCol(row, col) determinant() * sign
+					sign = -sign
+				}
+			}
+		}
 		this free(Owner Receiver)
 		result
 	}
 	adjugate: func -> This {
-		// TODO: At some point, when needed, implement for the general case.
-		version(safe) {
+		version(safe)
 			raise(!this take() isSquare, "Matrix must be square in FloatMatrix adjugate")
-			raise(this take() width != 3, "Adjugate implemented only for 3x3 matrices in FloatMatrix")
-		}
 		result := this take() cofactors() transpose()
 		this free(Owner Receiver)
 		result
 	}
 	determinant: func -> Float {
-		// TODO: At some point, when needed, implement for the general case.
 		t := this take()
-		version(safe) {
+		version(safe)
 			raise(!t isSquare, "Matrix must be square in FloatMatrix determinant")
-			raise(t width != 3, "Determinant implemented only for 3x3 matrices in FloatMatrix")
-		}
-		result := t[0, 0] * (t[1, 1] * t[2, 2] - t[2, 1] * t[1, 2]) -
+		result := 0.f
+		matrixOrder := t order
+		if (matrixOrder == 1)
+			result = t[0, 0]
+		else if (matrixOrder == 2)
+			result = t[0, 0] * t[1, 1] - t[0, 1] * t[1, 0]
+		else if (matrixOrder == 3)
+			result = t[0, 0] * (t[1, 1] * t[2, 2] - t[2, 1] * t[1, 2]) -
 			t[1, 0] * (t[0, 1] * t[2, 2] - t[2, 1] * t[0, 2]) +
 			t[2, 0] * (t[0, 1] * t[1, 2] - t[1, 1] * t[0, 2])
+		else {
+			sign := 1.f
+			for (column in 0 .. matrixOrder) {
+				submatrix := t _assembleWithoutRowCol(0, column) take()
+				result += t[0, column] * sign * submatrix determinant()
+				sign = -sign
+				submatrix free(Owner Sender)
+			}
+		}
 		this free(Owner Receiver)
 		result
 	}
+	_assembleWithoutRowCol: func (row, column: Int) -> This {
```

Comment:
Are we allowed to use abbreviations like `row, col` ?
Maybe not that important now, but simple optimization could be to add `submatrix` as third parameter to this function and let it fill in the values instead of allocating, this way you could avoid `N - 1` allocations when calculating determinant for the general case.
I mean this
```ooc
submatrix := FloatMatrix new(matrixOrder - 1, matrixOrder - 1)
for (column in 0 .. matrixOrder) {
    t _assembleWithoutRowCol(0, column, submatrix&)
    result += t[0, column] * sign * submatrix determinant()
....
```